File: bindings/nim/src/proven/safe_crypto.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Cryptographic safety operations with constant-time guarantees.
# Thin wrapper over libproven FFI -- all logic lives in Idris.

import std/options
import lib_proven

proc constantTimeCompare*(a, b: string): bool =
  ## Compare two strings in constant time to prevent timing attacks.
  if a.len == 0 and b.len == 0:
    return true
  let ptrA = if a.len > 0: unsafeAddr a[0] else: nil
  let ptrB = if b.len > 0: unsafeAddr b[0] else: nil
  let res = provenCryptoConstantTimeEq(ptrA, csize_t(a.len), ptrB, csize_t(b.len))
  if res.status == PROVEN_OK:
    return res.value
  false

proc constantTimeCompare*(a, b: openArray[byte]): bool =
  ## Compare two byte arrays in constant time.
  if a.len == 0 and b.len == 0:
    return true
  let ptrA = if a.len > 0: unsafeAddr a[0] else: nil
  let ptrB = if b.len > 0: unsafeAddr b[0] else: nil
  let res = provenCryptoConstantTimeEq(ptrA, csize_t(a.len), ptrB, csize_t(b.len))
  if res.status == PROVEN_OK:
    return res.value
  false

proc randomBytes*(n: int): Option[seq[byte]] =
  ## Generate cryptographically secure random bytes.
  ## Returns None if the system RNG fails.
  if n <= 0:
    return some(newSeq[byte](0))
  var buf = newSeq[byte](n)
  let status = provenCryptoRandomBytes(addr buf[0], csize_t(n))
  if status == PROVEN_OK:
    return some(buf)
  none(seq[byte])

proc randomHex*(n: int): Option[string] =
  ## Generate a cryptographically secure random hex string.
  ## Returns None if the system RNG fails.
  let bytes = randomBytes(n)
  if bytes.isNone:
    return none(string)
  const hexChars = "0123456789abcdef"
  var hex = newStringOfCap(n * 2)
  for b in bytes.get:
    hex.add hexChars[int(b shr 4)]
    hex.add hexChars[int(b and 0x0F)]
  some(hex)

File: bindings/lua/proven/safe_crypto.lua
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Cryptographic safety operations with constant-time guarantees.
--- @module proven.safe_crypto

local safe_crypto = {}

--- Compare two strings in constant time to prevent timing attacks.
--- Note: This is a best-effort implementation in pure Lua.
--- For true constant-time, use a C library binding.
--- @param a string
--- @param b string
--- @return boolean
function safe_crypto.constant_time_compare(a, b)
    if #a ~= #b then
        return false
    end

    if #a == 0 then
        return true
    end

    local result = 0
    for i = 1, #a do
        -- XOR each byte and OR into result
        result = bit32 and bit32.bor(result, bit32.bxor(a:byte(i), b:byte(i)))
            or (result | (a:byte(i) ~ b:byte(i)))  -- Lua 5.3+
    end

    return result == 0
end

--- Securely "zero" a string by creating a new zeroed string of the same length.
--- Note: Lua strings are immutable, so we cannot actually zero the original.
--- This returns a zeroed replacement string.
--- @param length number
--- @return string
function safe_crypto.secure_zero(length)
    return string.rep("\0", length)
end

--- Generate a table of random bytes.
--- Uses math.random which should be seeded properly for security.
--- For cryptographic use, prefer a proper CSPRNG binding.
--- @param n number
--- @return table Array of bytes
function safe_crypto.random_bytes(n)
    local bytes = {}
    for i = 1, n do
        bytes[i] = math.random(0, 255)
    end
    return bytes
end

--- Convert bytes table to string.
--- @param bytes table Array of bytes
--- @return string
function safe_crypto.bytes_to_string(bytes)
    local chars = {}
    for i, byte in ipairs(bytes) do
        chars[i] = string.char(byte)
    end
    return table.concat(chars)
end

--- Constant-time select: returns a if condition is true, b otherwise.
--- @param condition boolean
--- @param a any
--- @param b any
--- @return any
function safe_crypto.constant_time_select(condition, a, b)
    if condition then
        return a
    else
        return b
    end
end

return safe_crypto
